### ISSUE TITLE
Path computation unfavorable

### ISSUE THREAD

**Issue by HolQue** [test-fullautomation/robotframework-testsuitesmanagement#86]
Config\CConfig.py contains the following lines:

(1)
`if os.path.exists(self.__sNormalizePath(os.path.abspath('./' + sTestCfgDir))):`
(2)
`if os.path.exists(self.__sNormalizePath(os.path.abspath(sTestCfgDir))):`
(3)
```
if self.sTestCfgFile != '':
    self.sTestCfgFile = os.path.abspath(self.sTestCfgFile)
```
(4)
`if not os.path.isfile(self.__sNormalizePath(self.sTestCfgFile)):`
(5)
`oJsonCfgData = oJsonPreprocessor.jsonLoad(self.__sNormalizePath(self.sTestCfgFile))`
(6)
`sSchemaFile=str(pathlib.Path(__file__).parent.absolute() / "robot_schema.json")`

Here I have some doubts:

Are you really sure to need `'./'` in (1)? In (2) you do not use `'./'`.
The Python documentation tells about `abspath(path)`: Equivalent to `normpath(join(os.getcwd(), path))`.
Therefore I assume that `'./'` has no effect. Why do you make a difference here?

In my opinion something like `__sNormalizePath(os.path.abspath(path))` is unfavorable. When you introduce a function like `__sNormalizePath`, than this function should do the job completely. Why having the need to call an additional path function like `abspath` outside `__sNormalizePath`? Why not moving the call of `abspath` to inside `__sNormalizePath`?

(3) happens at the end - after you have created the path to `sTestCfgFile` under several conditions. But I wonder why you use only `os.path.abspath(...)` here. I would expect to use `__sNormalizePath(...)`.

(4) together with (5) is long winded. You should not use `__sNormalizePath(...)` everywhere you need the config file. You should use `__sNormalizePath(...)` only once with `self.sTestCfgFile` - and then use `self.sTestCfgFile` - without normalizing the same path again and again.

(6)
In your code you already use `os` functions like `os.path()` immediately. You also use `__sNormalizePath()`. With `pathlib.Path(...)` you introduce a further way of path computation. This mixture of different possibilities to do path computation is unfavorable and makes the code difficult to maintain. It's better to do things in a unique way.

Utils\LibListener.py contains the following line:
`localConfigFile = os.path.abspath(os.environ['ROBOT_LOCAL_CONFIG'])`
I am not really sure about all details of the code but also here I would assume that using `__sNormalizePath(...)` might be better than `os.path.abspath()`.

In general: Wherever you receive a path inside your code you should as early as possible use one common NormalizePath function.

**Comment by test-fullautomation** [comment]
Hi Holger,
Son uses the __sNormalizePath function from TML.
This function dosn't change absolut or relative state of the path.
Maybe it's useful to add an optional parameter
```
def .... "makeAbsolut=False" ...
```
 to avoid the use of abspath function every time.
Son: Please implement accordingly and call with  ```makeAbsolut=True``` if required.
Thank you,
 Thomas

**Comment by HolQue** [comment]
Hi Thomas,

I still have doubts. Even in case of a user provides a relative path for any input parameter of a script: Why do you want the script to continue with working with the relative version of the path? Instead of converting the relative path to an absolute one as soon as possible.

I cannot remember that I ever had a situation in which keeping a path relative was required inside my scripts.

Because keeping the paths from input parameters relative, is very dangerous. You run into danger that accidentally you use from any Python library a path function that does the conversion from relative to absolute paths implicitely. In this case you loose the control - because the current working directory is used as reference then.

The side effects are obvious. Example: You create a batch file to start a python script and pass a path as command line parameter.
* Batch file is located in same folder than the Python script => Python script runs properly (because of the same working directory)
* Batch file is located in another folder than the Python script => Python script fails (because now the working directory is the location of the batch file, and the Python script might not be prepared to handle this)

But to be more flexible, the parameter `makeAbsolut` might be useful. But I urgently propose to set the default value to `True`. Because in my opinion this is the standard use case - whereas `makeAbsolut=False` should be an exception.

Long time ago I also spend much effort to implement such a `NormalizePath` function, and I have had the same problem, but I solved the consideration of relative input paths in a little bit different way. Maybe it's worth to take a look at it:

https://github.com/test-fullautomation/python-extensions-collection/blob/develop/PythonExtensionsCollection/String/CString.py

**Comment by test-fullautomation** [comment]
Hi Holger,
it makes sense to keep the path as it was provided by the user when you need to log it after normalizing.
In this case it's annoying if you display an absolute path.
For all internal processing I agree that using an absolute path avoids many problems. Even debugging is easier.
```makeAbsolute=True``` as default is O.K. for me.
Thank you,
 Thomas

**Comment by HolQue** [comment]
Another example from current implementation of test suites management path computation.

Instead of `"[import]"   : "./exercise-05_config_common.json",`
I tried (just being snoopy):  `"[import]"   : "%CONFIG_COMMON%",`

With outcome:

`[ ERROR ] Loading of JSON configuration file failed! Reason: json file 'C:/RobotTest/projects/ROBFW_AIO/robotframework-tutorial/900_building_testsuites/exercise-05/config/exercise-05_config_default.json': 'File 'C:\RobotTest\projects\ROBFW_AIO\robotframework-tutorial\900_building_testsuites\exercise-05\config\%CONFIG_COMMON%' is not existing!'
`

Accidently the current working directory is added to the user input as prefix. This error message is wrong.

Should be something like:

`%CONFIG_COMMON% does not exist.`

**Comment by HolQue** [comment]
Hi Thomas,

I agree: In case we speak about a feedback to the user, the original input has to be part of the feedback - and not the already modified version of the user input. But this can be solved by another variable that stores this original content. No need to continue the further computation with the relative version of the path.

**Comment by HolQue** [comment]
Let' make it rough:
`"[import]"   : ".//..//config\\\\..\/config\\exercise-05_config_common.json",`
This is a valid path! And works in Windows.

In Linux the file is not found:
`Unknown exception occurs. Details: json file '/home/qth2hi/robotframework-tutorial/900_building_testsuites/exercise-05/config/exercise-05_config_default.json': 'File '/home/qth2hi/robotframework-tutorial/900_building_testsuites/exercise-05/config\\../config\exercise-05_config_common.json' is not existing!'`

But when I cross check this path in Linux with a stand alone Python test script like:

```
from PythonExtensionsCollection.String.CString import CString
sPath = r".//..//config\\\\..\/config\\exercise-05_config_common.json"
print(f"--- sPath: {sPath}")
sPathNormalized = CString.NormalizePath(sPath, sReferencePathAbs=__file__)
print(f"--- sPathNormalized: {sPathNormalized}")
if os.path.isfile(sPathNormalized):
   print("file does exist")
```

Then the result is:

```
--- sPath: .//..//config\\\\..\/config\\exercise-05_config_common.json
--- sPathNormalized: /home/qth2hi/test/config/exercise-05_config_common.json
file does exist
```
Outcome: The differences in path computations between Windows and Linux needs to be considered in more detail.

**Comment by test-fullautomation** [comment]
Hi Son,

This is a difficult topic :-)

Please copy Holger's path normalization function from CStrings and use is for every path calculation.
Please let Holger explain this function to you that you use it properly.

The same we do with all repositories where we need path normalizations.

Thank you,
 Thomas

**Comment by test-fullautomation** [comment]
Hi Son,

can you give me a status about this issue?

Task here was:

> Please copy Holger's path normalization function from CStrings and use is for every path calculation.
> Please let Holger explain this function to you that you use it properly.

Thank you, 
 Thomas

**Comment by HolQue** [comment]
Small addition: Please do not copy the code. The PythonExtensionCollection is available in every repository.

Access by:

`from PythonExtensionsCollection.String.CString import CString`

Like described in:

https://github.com/test-fullautomation/python-extensions-collection/blob/develop/PythonExtensionsCollection/PythonExtensionsCollection.pdf

**Comment by test-fullautomation** [comment]
Hi Holger,
I'm confused.
I don't want that every repository is dependent on PythonExtensionsCollection just because of a path calculation?
testsuites-management should have only jsonpreprocessor as dependency if installed from pip.
...
but while I write this I wonder if a dependency on PythonExtensionsCollection is really an issue when we have anyhow already a mandatory dependency to jsonpreprocessor. One dependency more doesn't change that we already need network access to pip to install testsuites-management.

@namsonx:
Conclusion: testsuites-management gets a new dependency PythonExtensionsCollection.
And we use the path normalizing function as Holger describes.

Thank you,
 Thomas

**Comment by HolQue** [comment]
Hi Thomas,

we already discussed this several times and the dependency is confirmed already. Every repository uses GenPackageDoc. And GenPackageDoc uses the PythonExtensionsCollection. Therefore we already have such dependencies in all repositories. They are listed already in the install dependencies:

`"INSTALLREQUIRES" : ["pypandoc","colorama","GenPackageDoc","PythonExtensionsCollection"],`

Therefore please no copy of code.

**Comment by test-fullautomation** [comment]
Hi Son,

what is the status of this issue using Holgers Proposal?

> Small addition: Please do not copy the code. The PythonExtensionCollection is available in every repository.
> 
> Access by:
> 
> `from PythonExtensionsCollection.String.CString import CString`
> 
> Like described in:
> 
> https://github.com/test-fullautomation/python-extensions-collection/blob/develop/PythonExtensionsCollection/PythonExtensionsCollection.pdf

Thank you,
 Thomas

**Comment by namsonx** [comment]
Hello Thomas,

I will apply the path normalizing function which Holger proposes.

Thank you,
Son

**Comment by test-fullautomation** [comment]
very good!

**Comment by namsonx** [comment]
Hello Thomas and Holger,

I created the pull request #159  to improve the path computation.

Thank you,
Son

**Comment by test-fullautomation** [comment]
solved and merged with #159
